backend/b2/b2.go

```diff
@@ -796,6 +804,22 @@ func (f *Fs) listBucketsToFn(fn listBucketFn) error {
 	return nil
 }
 
+// getbucketType finds the bucketType for the current bucket name
+// can be one of allPublic. allPrivate, or snapshot
+func (f *Fs) getbucketType() (bucketType string, err error) {
```

Comment:
`Does `listBuckets` work with limited keys? I don't think it does...

Maybe this should be integrated into getBucketID which is very likely to have been called already.  We could then cache type in the `Fs` in case it is called repeatedly

Reply:
with a Application Key limited to a bucket, `listBuckets` only list the bucket if the `bucketId` is included in the request which `listBucketsToFn` does.
if a key does not have the `listBuckets` permission at all then it wont work, same if the key does not have the `shareFiles` permission, however a default application key limited to a bucket has them.
i could not find a way to determine if a bucket was public or private without `listBuckets` 

when using the link command, as far as i can tell `getBucketID` is not called before `getBucketType`, however it might make sense to include `getBucketType` in `NewFs()` and cache the result.

please let me know what you think

Reply:
Good to know about the limited application key.

I think it would be a good idea to cache the bucket type.  If you wanted to add lots of links then using the API would be the way to do it, and if we don't cache the bucket type then we will be doing more calls to b2 than we need.

I would do something like this :-)

```diff
diff --git a/backend/b2/b2.go b/backend/b2/b2.go
index 65e643a11..499a417a3 100644
--- a/backend/b2/b2.go
+++ b/backend/b2/b2.go
@@ -162,8 +162,9 @@ type Fs struct {
 	bucket        string                       // the bucket we are working on
 	bucketOKMu    sync.Mutex                   // mutex to protect bucket OK
 	bucketOK      bool                         // true if we have created the bucket
-	bucketIDMutex sync.Mutex                   // mutex to protect _bucketID
+	bucketIDMutex sync.Mutex                   // mutex to protect _bucketID and _bucketType
 	_bucketID     string                       // the ID of the bucket we are working on
+	_bucketType   string                       // the Type of the bucket we are working on
 	info          api.AuthorizeAccountResponse // result of authorize call
 	uploadMu      sync.Mutex                   // lock for upload variable
 	uploads       []*api.GetUploadURLResponse  // result of get upload URL calls
```

And

```go
// _getBucketInfo finds the bucket ID and type for the current bucket
// name
//
// must be called with the bucketIDMutex held
func (f *Fs) _getBucketInfo() (bucketID, bucketType string, err error) {
	if f._bucketID != "" {
		return f._bucketID, f._bucketType, nil
	}
	var foundBucket *api.Bucket
	err = f.listBucketsToFn(func(bucket *api.Bucket) error {
		if bucket.Name == f.bucket {
			foundBucket = bucket
		}
		return nil

	})
	if foundBucket == nil {
		return "", "", fs.ErrorDirNotFound
	}
	f._bucketID = foundBucket.ID
	f._bucketType = foundBucket.Type
	return f._bucketID, f._bucketType, nil
}

// getBucketID finds the ID for the current bucket name
func (f *Fs) getBucketID() (bucketID string, err error) {
	f.bucketIDMutex.Lock()
	defer f.bucketIDMutex.Unlock()
	bucketID, _, err = f._getBucketInfo()
	return bucketID, err
}

// getBucketType finds the type for the current bucket name
func (f *Fs) getBucketType() (bucketType string, err error) {
	f.bucketIDMutex.Lock()
	defer f.bucketIDMutex.Unlock()
	_, bucketType, err = f._getBucketInfo()
	return bucketType, err
}
```

Reply:
i agree, however the `bucketID` is included in the response from `AuthorizeAccount` and therefore when using a single bucket ApplicationKey `getBucketID` does not make any call, if we include `getBucketType` (that is **not** included in the response from `AuthorizeAccount`) in one call it will require that every command will add a call to `listBuckets`, even ones that until now did not require it
therefore i think that while caching it, it should be kept as a separate function from `getBucketID`

if you agree i will finish this up.

Reply:
Ah I see what you mean.  Let's do what you suggest :-)

Reply:
done in the latest